## CSC4815 Machine Learning
Homework 2 due on 3/2 at midnight

__Noah Foilb__

Rules:
- Solve each problem only in the given cell and show the final result. Do not add cells.
- You are welcome to solve a problem incrementally. However, do not output any debugging information in your final solution.
- A solution ending with a syntax or runtime error will get zero points no matter how much you worked on it. It will be much better to submit an error-free partial solution than a solution with an error.
- Each problem can be solved in different ways. However, you may use pandas/DataFrame methods and vanilla Python operations only for this homework. You may not import any external modules.
- Always try to make the output informative and intuitive. That's what your client will care about in the end.
- Optional Questions will not be graded. They are for those who are interested in going extra miles.
- The accompanying pandas cheatsheet includes useful information for this assignment.

Execute the following cell to ensure you are using Python 3 or above.

In [1]:
!python --version

Python 3.8.3


# Problem 1 (Missing Values)

(4 points) Solve this problem in the following sequence:

0. Load the titanic_.csv file. The column delimiter used in the file is not a comma.
1. Show the number of missing values of each column.
2. Show the missing value rate with respect to the number of rows of the table.
3. Show the column names whose missing value rate (MVR) is greater than 70%.
4. Show the schema of the table after the columns in Step 3 are removed (in other words, retain columns whose MVR is less than or equal to 70%).
5. Show the mean and median of the age column. Fill the missing values in the column with 0 and show the new mean. Fill the missing values with the median of the orginal column this time and show the new mean. Finally fill the missing values with the mean of the orginal column and show the new mean.

In [2]:
# Step 0:
import pandas as pd                                    # Import pandas

df = pd.read_csv("titanic_.csv",delimiter='|')         # Read_csv to get data into python and delimiter to create dataframe
df.head()                                              # Display Results

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

       age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.0000      0      0   24160  211.3375       B5        S    2    NaN   
1   0.9167      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.0000      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

In [3]:
# Step 1:
print("The number of missing values in each column:","\n",
      df.isna().sum())                                             # Print the number of missing values in each column 

The number of missing values in each column: 
 pclass          0
survived        0
name            0
sex             0
age           263
sibsp           0
parch           0
ticket          0
fare            1
cabin        1014
embarked        2
boat          823
body         1188
home.dest     564
dtype: int64


In [4]:
# Step 2:
MVR = round(sum(df.isna().sum())/len(df),2)                      # Calculate the MVR 
print("There is an MVR", MVR)                                    # Display Results


There is an MVR 2.94


In [5]:
# Step 3:
print("The two columns that have MVR over 70% are:")             # Print the starting line

for i in range(len(df.columns)):                                 # Make for loop to go through the length of all columns
    if (df.iloc[:,i].isna().sum())/len(df)*100 > 70:             # Requirement that the only columns that are getting de are <70
        print(df.columns[i])                                     # Print the columns that need to be deleted



The two columns that have MVR over 70% are:
cabin
body


In [6]:
# Step 4:
for i in range(len(df.columns)):                               # Same setup as before except the range is backwards
    if (df.iloc[:,i].isna().sum())/len(df)*100 > 70: 
        print(df.columns[i])
        df.drop(df.columns[i],axis = 1, inplace = True)         # Delete the columns
        
df.head()    

cabin
body


IndexError: single positional indexer is out-of-bounds

In [ ]:
# Step 5:
original_mean = round(df["age"].mean(),4)                                   # Calculate the original mean
original_median = round(df["age"].median(),1)                               # Calculate the original median
original_sum = (df.iloc[:,4].isna().sum())                                  # Calculate the original sum of Na's

zero_mean = round(df["age"].fillna(0).mean(),4)                             # Mean where nans are zeros
 
median_mean = round(df["age"].fillna(original_median).mean(),4)             # Mean where nans are medians
 
mean_mean = round(df["age"].fillna(original_mean).mean(),4)                 # Mean where nans are means

print("Original sum =" + str(original_sum) + ", median=" + str(original_median) + ", mean=" + str(original_mean),"\n" +
     "Mean after filling in with zero=" + str(zero_mean), "\n" +
     "Mean after filling in with median=" + str(median_mean), "\n" +                # Print Results
     "Mean after filling in with mean=" + str(mean_mean))


Correct output:

> `Original sum=263, median=28.0, mean=29.8811`<br/>
> `Mean after filling in with zero=23.8775`<br/>
> `Mean after filling in with median=29.5032`<br/>
> `Mean after filling in with mean=29.8811`<br/>


(Optional Question) Why is a new mean greater or smaller than the original?

# Problem 2 (Outlier Detection)

(2 points) Review the revised slides on outliers in the first Lecture Notes first. Solve this problem in the following sequence:

0. Load the California housing data (housing.csv).
1. Create a copy of the total_bedrooms column as a DataFrame. Calculate the z-score of the total_bedrooms and add it to the new DataFrame as a new column. Show the schema of the new DataFrame.
2. Show the sum and mean of the z_score column.
3. Show the mean value of total_bedrooms, the number of z-score values which are greater than 5 and the mean of the corresponding total_bedrooms.

In [ ]:
# Step 0:
# Common imports
import os
import pandas as pd

DOWNLOAD_ROOT = "https://raw.githubusercontent.com/ageron/handson-ml2/master/"
HOUSING_PATH = os.path.join("datasets", "housing", "housing.csv")

if os.path.exists(HOUSING_PATH):
    print(f"Loading {HOUSING_PATH} locally...")
    housing = pd.read_csv(HOUSING_PATH)
else:
    print(f"Downloading {HOUSING_PATH} online...")
    housing = pd.read_csv('https://raw.githubusercontent.com/ageron/handson-ml2/master/datasets/housing/housing.csv')

In [ ]:
# Step 1:

Bedroom = housing[['total_bedrooms']]                      # set the total bedrooms into their own datafram 
Bedroom['z_score'] = (Bedroom['total_bedrooms']-Bedroom['total_bedrooms'].mean())/Bedroom['total_bedrooms'].std()   #make zscore
Bedroom.head(10)                                           # Display results

In [ ]:
# Step 2:

Bedroom_mean = float("{:.15f}".format(Bedroom['z_score'].mean()))                         # Find the mean of bedroom zscore
Bedroom_sum = float("{:.15f}".format(Bedroom['z_score'].sum()))                           # Find sum of zscores

print('Z-score column: sum=' + str(Bedroom_sum) + ', mean=' + str(Bedroom_mean) )

Correct output when 15 digits are taken after the decimal point:

> `Z-score column: sum=-0.000000000001535, mean=-0.000000000000000`

(Optional Question) Explain why both sum and mean of z-scores are close to zero.

In [ ]:
# Step 3:

Z = len(Bedroom[Bedroom['z_score']>5])                                        # Number of Z scores
mean_bedroom = round(Bedroom['total_bedrooms'].mean(),1)                      # Mean total bedrooms
num_z = round(Bedroom[Bedroom['z_score']>5]['total_bedrooms'].mean(),1)       # Mean total bedrooms > 5 z score

print('Mean total_bedrooms =' + str(mean_bedroom) + ',numer of z-scores >5=' 
      + str(Z) + ',their mean total bedrooms=' + str(num_z))                  # Display results

Corret output when 1 digit after decimal point is taken:

> `Mean total_bedrooms=537.9,number of z-scores > 5=109,their mean total bedrooms=3489.6`

(Optional Question) What can you tell about the total_bedrooms column? Does the column have outliers or not?

# Problem 3 (Discretization and Binirization)

(2 points) Solve this problem in the following sequence:

0. Load the titanic_.csv file. The column delimiter used in the file is not a comma.
1. Draw a bar graph of the age column with no attribute tranformation.
2. Draw a histogram of the same column using 8 equiwidth bins.
3. Perform one-hot encoding over the embarked column, add the new one-hot encoded columns to the dataframe, and show the embarked and the one-hot encoded column values of row indexes 0, 9, and 684.
4. Show the one-hot encoding of every null value.

In [ ]:
# Step 0:
import pandas as pd                                    # Import pandas

df = pd.read_csv("titanic_.csv",delimiter='|')         # Read_csv to get data into python and delimiter to create dataframe
df.head()    

In [ ]:
# Step 1:
Age = df['age'].round()
Age.plot.bar(figsize=(10,8))                  # barplot

# So I know I can make a better bar graph but you said to not do any attribute transformation so I wasnt sure if the excersize 
# Here was to see how much better the histogram was compared to the bar plot. Just for clarification I realize that 
# some ages are not whole numbers so If i were to make this graph better I would include a round function to the age column
# and display that bar plot

(Optional Question) What can you tell from the plot concerning the ages of the passangers?

In [ ]:
# Step 2:
df["age"].hist(bins=8)                                        # Display histogram with 8 equal bin


(Optional Question) Can you say more about the ages of the passangers?

In [ ]:
# Step 3:
Df = pd.get_dummies(df['embarked'], prefix='embarked')                    # Create the one hot encoding
DF = pd.concat([df,Df],axis = 1)                                          # Join together the two datasets
DF[DF.columns[[10,14,15,16]]]                                             # Only display the columns we want

In [ ]:
# Step 4:
dF = DF[DF.columns[[10,14,15,16]]]                                      # Only want o use the columns we want
dF[dF.isna().any(axis=1)]                                               # Display the columns with nans in embarked

# Problem 4 (Correlation)

(2 points) Solve this problem in the following sequence:

0. Load the titanic_.csv file. The column delimiter used in the file is not a comma.
1. Show the correlation value of the column which is most correlated to the survive column.

In [ ]:
# Step 0:
import pandas as pd                                    # Import pandas

df = pd.read_csv("titanic_.csv",delimiter='|')         # Read_csv to get data into python and delimiter to create dataframe
df.head()    

In [ ]:
# Step 1:
corr_matrix = df.corr()                                                # Find the correleation matix
corr_matrix["survived"].sort_values(ascending=False).head(2)           # Sort the values by survived and display the top result

Correct output:

> `survived    1.000000`<br/>
> `fare        0.244265`<br/>
> `Name: survived, dtype: float64`</br>

(Optional Questions) Which column or a set of columns could the best indicator for the survived passangers?